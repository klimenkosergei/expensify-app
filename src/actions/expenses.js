import database from '../firebase/firebase';

const addExpense = expense => ({
  type: 'ADD_EXPENSE',
  expense
});

const startAddExpense = ({
  description = '',
  note = '',
  amount = 0,
  createdAt = 0
} = {}) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    const expense = { description, note, amount, createdAt };

    return database
      .ref(`users/${uid}/expenses`)
      .push(expense)
      .then(ref => {
        dispatch(
          addExpense({
            id: ref.key,
            ...expense
          })
        );
      });
  };
};

const removeExpense = id => ({
  type: 'REMOVE_EXPENSE',
  id
});

const startRemoveExpense = id => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    // Return promise so that we can chain .then in test case
    return database
      .ref(`users/${uid}/expenses/${id}`)
      .remove()
      .then(() => {
        dispatch(removeExpense(id));
      });
  };
};

const editExpense = (id, updates) => ({
  type: 'EDIT_EXPENSE',
  id,
  updates
});

const startEditExpense = (id, updates) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database
      .ref(`users/${uid}/expenses/${id}`)
      .set({
        ...updates
      })
      .then(() => {
        dispatch(editExpense(id, updates));
      });
  };
};

const setExpenses = expenses => ({
  type: 'SET_EXPENSES',
  expenses
});

const startSetExpenses = () => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    const expenses = [];

    return database
      .ref(`users/${uid}/expenses`)
      .once('value')
      .then(snapshot => {
        snapshot.forEach(childSnapshot => {
          expenses.push({
            id: childSnapshot.key,
            ...childSnapshot.val()
          });
        });
        dispatch(setExpenses(expenses));
      });
  };
};

export {
  startAddExpense,
  addExpense,
  removeExpense,
  startRemoveExpense,
  editExpense,
  startEditExpense,
  setExpenses,
  startSetExpenses
};
