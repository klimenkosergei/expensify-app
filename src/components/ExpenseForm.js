import React from 'react';
import moment from 'moment';
import 'react-dates/initialize';
import { SingleDatePicker } from 'react-dates';

export default class ExpenseForm extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      description: props.expense ? props.expense.description : '',
      amount: props.expense ? (props.expense.amount / 100).toString() : '',
      note: props.expense ? props.expense.note : '',
      createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
      calendarFocused: false,
      error: ''
    };
  }
  onDescriptionChange = e => {
    const description = e.target.value;
    this.setState(prevState => ({ description }));
  };
  onAmountChange = e => {
    const amount = e.target.value;
    if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
      this.setState(prevState => ({ amount }));
    }
  };
  onNoteChange = e => {
    const note = e.target.value;
    this.setState(prevState => ({ note }));
  };
  onDateChange = date => {
    if (date) {
      this.setState(prevState => ({ createdAt: date }));
    }
  };
  onFocusChange = ({ focused }) => {
    this.setState(prevState => ({ calendarFocused: focused }));
  };
  onSubmit = e => {
    e.preventDefault();
    if (!this.state.description || !this.state.amount) {
      this.setState(prevState => ({
        error: 'Please provide description and amount.'
      }));
    } else {
      // Clear error state
      this.setState(prevState => ({ error: '' }));
      this.props.onSubmit({
        description: this.state.description,
        note: this.state.note,
        amount: parseFloat(this.state.amount) * 100, // Times 100 since our format is in pennies
        createdAt: this.state.createdAt.valueOf()
      });
    }
  };
  render() {
    return (
      <div>
        {this.state.error && <p>{this.state.error}</p>}
        <form onSubmit={this.onSubmit} className='form'>
          <input
            type='text'
            placeholder='Description'
            autoFocus
            value={this.state.description}
            onChange={this.onDescriptionChange}
            className='form__input'
          />
          <input
            type='text'
            placeholder='Amount'
            value={this.state.amount}
            onChange={this.onAmountChange}
            className='form__input'
          />
          <SingleDatePicker
            date={this.state.createdAt}
            onDateChange={this.onDateChange}
            focused={this.state.calendarFocused}
            onFocusChange={this.onFocusChange}
            numberOfMonths={1}
            displayFormat={() => 'DD/MM/Y'}
            firstDayOfWeek={1}
            hideKeyboardShortcutsPanel={true}
            isOutsideRange={() => false}
          />
          <textarea
            placeholder='Add a note to expense (optional)'
            value={this.state.note}
            onChange={this.onNoteChange}
            rows='4'
            className='form__textarea'
          ></textarea>
          <button type='submit' className='form__submit'>
            Submit
          </button>
        </form>
      </div>
    );
  }
}
