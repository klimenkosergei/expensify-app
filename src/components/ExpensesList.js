import React from 'react';
import { connect } from 'react-redux';

import getVisibleExpenses from '../selectors/expenses';
import ExpensesListItem from './ExpensesListItem';

const ExpensesList = props => (
  <div className='list'>
    <div className='list__wrapper'>
      <div className='list__header'>
        <span className='list__header-item'>Expenses</span>
        <span className='list__header-item'>Amount</span>
      </div>
      {props.expenses.length !== 0 ? (
        props.expenses.map(expense => (
          <ExpensesListItem key={expense.id} {...expense} />
        ))
      ) : (
        <p className='list__item list__item--empty'>No expenses found</p>
      )}
    </div>
  </div>
);

const mapStateToProps = state => ({
  expenses: getVisibleExpenses(state.expenses, state.filters)
});

const ConnectedExpensesList = connect(mapStateToProps)(ExpensesList);

export { ExpensesList, ConnectedExpensesList as default };
