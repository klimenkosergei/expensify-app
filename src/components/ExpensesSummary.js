import React from 'react';
import { connect } from 'react-redux';
import numeral from 'numeral';

import getVisibleExpenses from '../selectors/expenses';
import getTotalExpenses from '../selectors/expenses-total';

const ExpensesSummary = props => (
  <div className='summary__wrapper'>
    <p className='summary__text'>
      Viewing{' '}
      <span className='summary__text summary__text--bold'>
        {props.expensesAmount}
      </span>{' '}
      expenses totalling{' '}
      <span className='summary__text summary__text--bold'>
        {numeral(props.expensesValue / 100).format('$0,0.00')}
      </span>
    </p>
  </div>
);

const mapStateToPros = state => {
  const expenses = getVisibleExpenses(state.expenses, state.filters);
  const expensesValue = getTotalExpenses(expenses);

  return {
    expensesAmount: expenses.length,
    expensesValue
  };
};

const ConnectedExpensesSummary = connect(mapStateToPros)(ExpensesSummary);

export { ExpensesSummary, ConnectedExpensesSummary as default };
