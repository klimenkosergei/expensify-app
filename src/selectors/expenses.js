import moment from 'moment';

const getVisibleExpenses = (expenses, { text, sortBy, startDate, endDate }) =>
  expenses
    .filter(expense => {
      const createdAt = moment(expense.createdAt);
      const textMatch = expense.description
        .toLowerCase()
        .includes(text.toLowerCase());
      const startDateMatch = startDate
        ? startDate.isSameOrBefore(createdAt, 'day')
        : true;
      const endDateMatch = endDate
        ? endDate.isSameOrAfter(createdAt, 'day')
        : true;

      return textMatch && startDateMatch && endDateMatch;
    })
    .sort((a, b) => {
      // Chain sorting to a filtered array
      switch (sortBy) {
        case 'date':
          return a.createdAt > b.createdAt ? -1 : 1;
        case 'amount':
          return a.amount > b.amount ? -1 : 1;
        default:
          return;
      }
    });

export default getVisibleExpenses;
