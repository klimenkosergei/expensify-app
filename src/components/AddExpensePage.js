import React from 'react';
import { connect } from 'react-redux';

import { startAddExpense } from '../actions/expenses';
import ExpenseForm from './ExpenseForm';

class AddExpensePage extends React.Component {
  constructor(props) {
    super(props);
  }
  addExpense = expense => {
    this.props.addExpense(expense);
    this.props.history.push('/projects/expensify');
  };
  render() {
    return (
      <div className='create'>
        <div className='create__wrapper'>
          <h2 className='create__title'>Create Expense</h2>
          <ExpenseForm onSubmit={this.addExpense} />
        </div>
      </div>
    );
  }
}

const mapDispatchToProps = dispatch => ({
  addExpense: expense => {
    dispatch(startAddExpense(expense));
  }
});

const ConnectedAddExpensePage = connect(
  null,
  mapDispatchToProps
)(AddExpensePage); // Pass null for the mapStateToProps

export { AddExpensePage, ConnectedAddExpensePage as default };
