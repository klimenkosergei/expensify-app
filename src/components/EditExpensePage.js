import React from 'react';
import { connect } from 'react-redux';

import { startEditExpense, startRemoveExpense } from '../actions/expenses';
import ExpenseForm from './ExpenseForm';

class EditExpensePage extends React.Component {
  constructor(props) {
    super(props);
  }
  editExpense = expense => {
    this.props.editExpense(this.props.expense.id, expense);
    this.props.history.push('/projects/expensify');
  };
  removeExpense = () => {
    this.props.removeExpense(this.props.expense.id);
    this.props.history.push('/projects/expensify');
  };
  render() {
    return (
      <div className='create'>
        <div className='create__wrapper'>
          <h2 className='create__title'>Edit Expense</h2>
          <ExpenseForm
            expense={this.props.expense}
            onSubmit={this.editExpense}
          />
          <button onClick={this.removeExpense} className='create__btn'>
            Delete
          </button>
        </div>
      </div>
    );
  }
}

const mapStateToProps = (state, props) => ({
  expense: state.expenses.find(expense => expense.id === props.match.params.id)
});

const mapDispatchToProps = dispatch => ({
  editExpense: (expenseId, expense) => {
    dispatch(startEditExpense(expenseId, expense));
  },
  removeExpense: expenseId => {
    dispatch(startRemoveExpense(expenseId));
  }
});

const ConnectedEditExpensePage = connect(
  mapStateToProps,
  mapDispatchToProps
)(EditExpensePage);

export { EditExpensePage, ConnectedEditExpensePage as default };
