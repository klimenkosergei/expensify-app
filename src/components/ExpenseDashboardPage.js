import React from 'react';
import ExpensesList from './ExpensesList';
import ExpensesSummary from './ExpensesSummary';
import ExpensesListFilters from './ExpensesListFilters';

const ExpenseDashboardPage = () => (
  <div>
    <ExpensesSummary />
    <ExpensesListFilters />
    <ExpensesList />
  </div>
);

export default ExpenseDashboardPage;
