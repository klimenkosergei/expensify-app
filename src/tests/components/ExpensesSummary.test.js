import React from 'react';
import { shallow } from 'enzyme';

import { ExpensesSummary } from '../../components/ExpensesSummary';
import getTotalExpenses from '../../selectors/expenses-total';
import expenses from '../fixtures/expenses';

test('should render ExpensesSummary with one visible expense', () => {
  const expensesValue = getTotalExpenses([expenses[0]]);
  const wrapper = shallow(
    <ExpensesSummary
      expensesAmount={[expenses[0]].length}
      expensesValue={expensesValue}
    />
  );
  expect(wrapper).toMatchSnapshot();
});

test('should render ExpensesSummary with multiple visible expense', () => {
  const expensesValue = getTotalExpenses(expenses);
  const wrapper = shallow(
    <ExpensesSummary
      expensesAmount={expenses.length}
      expensesValue={expensesValue}
    />
  );
  expect(wrapper).toMatchSnapshot();
});
