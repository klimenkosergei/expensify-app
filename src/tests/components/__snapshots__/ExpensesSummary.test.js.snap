// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpensesSummary with multiple visible expense 1`] = `
<div
  className="summary__wrapper"
>
  <p
    className="summary__text"
  >
    Viewing
     
    <span
      className="summary__text summary__text--bold"
    >
      3
    </span>
     
    expenses totalling
     
    <span
      className="summary__text summary__text--bold"
    >
      $1,102.95
    </span>
  </p>
</div>
`;

exports[`should render ExpensesSummary with one visible expense 1`] = `
<div
  className="summary__wrapper"
>
  <p
    className="summary__text"
  >
    Viewing
     
    <span
      className="summary__text summary__text--bold"
    >
      1
    </span>
     
    expenses totalling
     
    <span
      className="summary__text summary__text--bold"
    >
      $4.00
    </span>
  </p>
</div>
`;
